diesel/src/connection/transaction_manager.rs

```diff
@@ -114,43 +115,57 @@ where
         let transaction_depth = conn.transaction_state().transaction_depth;
         if transaction_depth <= 1 {
             match conn.batch_execute("COMMIT") {
-                // When any of these kinds of error happen on `COMMIT`, it is expected
-                // that a `ROLLBACK` would succeed, leaving the transaction in a non-broken state.
-                // If there are other such errors, it is fine to add them here.
-                e @ Err(Error::DatabaseError(DatabaseErrorKind::SerializationFailure, _))
-                | e @ Err(Error::DatabaseError(DatabaseErrorKind::ReadOnlyTransaction, _)) => {
-                    let r = conn.batch_execute("ROLLBACK");
-                    conn.transaction_state().change_transaction_depth(-1, r)?;
-                    e
+                Ok(()) => {
+                    let _r = conn
+                        .transaction_state()
+                        .change_transaction_depth(-1, Ok(()));
+                    Ok(())
+                }
+                Err(error) => {
+                    let error = match conn.process_commit_error(transaction_depth, error) {
+                        CommitErrorOutcome::RollbackAndThrow(error) => {
+                            let r = conn.batch_execute("ROLLBACK");
+                            conn.transaction_state().change_transaction_depth(-1, r)?;
+                            error
+                        }
+                        CommitErrorOutcome::Throw(error) => {
+                            let _r = conn
+                                .transaction_state()
+                                .change_transaction_depth(-1, Ok(()));
+                            error
+                        }
```

Comment:
I think you make some good points wrt decrementing the transaction depth upon failure.

I'm less sanguine about the variant names you suggest. IMO, the role of the processor should really be to specify the action to take following the commit failure, because this depends so much on the database implementation, but I'm open to changing these if you make a good argument.

> When implementor of CommitErrorProcessor cannot guarantee that TransactionAlreadyRolledBack is true, then it shall return ConnectionBroken. This guarantees that we don't ever release a connection to the pool with transaction_depth=0 when that is in fact not true at the database level.

Well, returning `ConnectionBroken` / `TransactionStatusUnknown` doesn't guarantee anything of the sort, unfortunately, since we don't have a switch at the moment to turn the connection into a broken connection. We can return `DatabaseErrorKind::ClosedConnection` from the manager, but this is not going to do anything about the connection. It's tricky to get right: what if a particular connection type could recover from a transport error (I don't think Postgres can).

> With regards to process_commit_error's default impl, I think we should either:
>
> * Remove the default impl completely

I could go along with it as long as we have a utility function which is called for the mysql and sqlite implementations, since I don't believe they need to be different at this time (AFAIK, we don't have anything to test the status of the connection in either case)

>  For the same reason, I think the impl for postgres should not do _ => CommitErrorOutcome::TransactionAlreadyRolledBack either, and should instead return that only for specified errors, or check transaction_status if it doesn't know.

I'd be happy to document that, but my understanding is that as long as a deferred trigger fired upon commit (and failed), any kind of exception could have been thrown by the trigger (and automatically rollback the connection). That said, enumerating the errors in the match rather than doing a default match would be safer.

> check transaction_status if it doesn't know

That's not very useful for if the transaction depth is > 1.

Reply:
> Well, returning `ConnectionBroken` doesn't guarantee anything of the sort, unfortunately, since we don't have a switch at the moment to turn the connection into a broken connection

What I'm trying to guarantee is that "we don't ever release a connection to the pool **with `transaction_depth`=0** when that is in fact not true".
What I mean by that is that specifically, i'd like to maintain the behavior that if we're unsure, we leave the transaction depth higher, so that the caller can take additional action with the knowledge that stuff was not properly closed.

IIUC, this property was here before, and wouldn't be if we don't introduce the option to throw and not decrement transaction depth.

Reply:
Yeah, but thinking about it, and after some diving into Postgres' code (the only database we need to be concerned about wrt broken connections/transactions since we don't/can't check this for Mysql/Sqlite), connections do **NOT** appear to be able to recover from a `CONNECTION_BAD` status (unless you reset them by reopening them). So, operationally, once the connection to the Postgres server is lost, the transaction depth of the manager is irrelevant, since any future operation will fail with an error.

Now, from a correctness point of view, there is no telling the state of the server, since the connection may have been lost before or after the commit, which may or may not have processed correctly. The correct thing to do, IMO, would be to introduce something like 

```rust
enum TransactionManagerStatus {
   Valid { transaction_depth: i32 } // Not sure that I like an `i32` that `get_transaction_depth()` returns as an u32, but that's not the point here....
   Invalid // Hosed, you can't play with transactions with this connection again
}
```

> What I mean by that is that specifically, i'd like to maintain the behavior that if we're unsure, we leave the transaction depth higher, so that the caller can take additional action with the 
knowledge that stuff was not properly closed.

But that's misleading as well. IMO, the most robust behaviour in case of lost connection would to break the transaction manager and return a specific error wrapping the original cause, but at least for the "specific error" part, I intended to submit that in a specific PR.

Reply:
@weiznich , any thought regarding changing the API of transaction manager to potentially make it "broken" in case we get `PQTRANS_UNKNOWN` from PG (or equivalent for Mysql/Sqlite)?

Reply:
@Emm I hadn't the time to look at this yet. Will try to do that this week.

Reply:
Changing the transaction data from a simple integer that tracks only the depth to an enum that tracks also if the connection contains a broken connection is a great ideas. Given that we could easily implement behavior that just returns an error from any called function as soon as the connection contains a broken connection. This should drastically improve the behavior around panicing in transactions in combination with the usage of connection pools. So :+1: from me here. 
(Its totally fine to implement not all of that in this pr.)

Reply:
All right. I'll give a go at updating the transaction manager in this PR, changing to a "broken" transaction manager, based on getting `PQTRANS_UNKNOWN`. We can implement a similar mechanism to implement a broken state for the connection based on getting `CONNECTION_BAD` at the connection level, which is important if people get a broken connection outside of the transaction manager.

Thanks for your feedback